apps/heft/src/plugins/SassTypingsPlugin/SassTypingsGenerator.ts

```diff
@@ -0,0 +1,96 @@
+import * as path from 'path';
+import { render, Result } from 'node-sass';
+import * as postcss from 'postcss';
+import * as cssModules from 'postcss-modules';
+import { StringValuesTypingsGenerator, IStringValueTypings } from '@rushstack/typings-generator';
+import { LegacyAdapters } from '@rushstack/node-core-library';
+
+/**
+ * @public
+ */
+export interface ISassTypingsGeneratorOptions {
+  buildFolder: string;
+}
+
+/**
+ * @internal
+ */
+interface IClassMap {
+  [className: string]: string;
+}
+
+/**
+ * Generates type files (.d.ts) for Sass/SCSS/CSS files.
+ *
+ * @public
+ */
+export class SassTypingsGenerator extends StringValuesTypingsGenerator {
+  private _classMap: IClassMap;
+
+  /**
+   * @param buildFolder - The project folder to search for Sass files and generate typings.
+   */
+  public constructor(options: ISassTypingsGeneratorOptions) {
+    const srcFolder: string = path.join(options.buildFolder, 'src');
+    const generatedTsFolder: string = path.join(options.buildFolder, 'temp', 'sass-ts');
+    super({
+      srcFolder,
+      generatedTsFolder,
+      exportAsDefault: true,
+      exportAsInterfaceName: 'IExportStyles',
+      fileExtensions: ['scss', 'sass', 'css'],
+      parseAndGenerateTypings: async (fileContents: string, filePath: string) => {
+        const css: string = await this._transpileSassAsync(fileContents, filePath, options.buildFolder);
+        const classNames: string[] = await this._getClassNamesFromCSSAsync(css, filePath);
+        const sortedClassNames: string[] = classNames.sort((a, b) => a.localeCompare(b));
+        const sassTypings: IStringValueTypings = {
+          typings: []
+        };
+        for (const exportName of sortedClassNames) {
+          sassTypings.typings.push({ exportName });
+        }
+
+        return sassTypings;
+      }
+    });
+  }
+
+  private async _transpileSassAsync(
+    fileContents: string,
+    filePath: string,
+    buildFolder: string
+  ): Promise<string> {
+    const result: Result = await LegacyAdapters.convertCallbackToPromise(render, {
+      data: fileContents,
+      importer: (url: string) => ({ file: this._patchSassUrl(url) }),
+      includePaths: [path.join(buildFolder, 'node_modules'), path.join(buildFolder, 'src')],
+      indentedSyntax: path.extname(filePath).toLowerCase() === '.sass'
+    });
+
+    return result.css.toString();
+  }
+
+  private _patchSassUrl(url: string): string {
+    if (url[0] === '~') {
+      url = 'node_modules/' + url.substr(1);
+    } else if (url === 'stdin') {
```

Comment:
When/where does this occur out of curiosity?

Reply:
Good question! 

There is an example in `build-tests/heft-sass-test/src/styles.sass` that you can refer to.

In Sass, it is common to abbreviate 3rd party package imports with `~`, and something magically resolves it to "./node_modules/". For example, to import Sass mixins from the buttono project, it is typically done like so:
`@import '~buttono/buttono'`

For some reason, even though this is standard, it doesn't seem to resolve the import without this helper method.

Reply:
(BTW. I stole this method from gulp-core-build-sass).

Reply:
#Sorry, I should clarify. When does the 'stdin' value get used? The tilde makes sense

Reply:
Oh! I have absolutely no idea. I copied that from gulp-core-build-sass. 

I can't think of a scenario, so I should probably just remove it.

Reply:
Ok. It works great without the stdin statement. So I've removed it.